external/storm-opentsdb/src/main/java/org/apache/stom/opentsdb/bolt/OpenTsdbBolt.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    /**
+     * When it is invoked, this bolt acks only the tuples which have successful metrics stored into OpenTSDB and fails
+     * the respective tuples of the failed metrics.
+     *
+     * @return same instance by setting {@code failTupleForFailedMetrics} to true
+     */
+    public OpenTsdbBolt failTupleForFailedMetrics() {
+        this.failTupleForFailedMetrics = true;
+        return this;
+    }
+
+    @Override
+    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+        this.collector = collector;
+        batchHelper = new BatchHelper(batchSize, collector);
+        openTsdbClient = openTsdbClientBuilder.build();
+    }
+
+    @Override
+    public void execute(Tuple tuple) {
+        try {
+            if (batchHelper.shouldHandle(tuple)) {
+                final List<OpenTsdbMetricDatapoint> metricDataPoints = getMetricPoints(tuple);
+                for (OpenTsdbMetricDatapoint metricDataPoint : metricDataPoints) {
+                    metricPointsWithTuple.put(metricDataPoint, tuple);
+                }
+                batchHelper.addBatch(tuple);
+            }
+
+            if (batchHelper.shouldFlush()) {
+                LOG.debug("Sending metrics of size [{}]", metricPointsWithTuple.size());
+
+                ClientResponse.Details clientResponse = openTsdbClient.writeMetricPoints(metricPointsWithTuple.keySet());
+
+                if(failTupleForFailedMetrics && clientResponse.getFailed() > 0) {
+                    final List<ClientResponse.Details.Error> errors = clientResponse.getErrors();
+                    LOG.error("Some of the metric points failed with errors: [{}]", clientResponse);
+                    if(errors != null && !errors.isEmpty()) {
+
+                        Set<Tuple> failedTuples = new HashSet<>();
+                        for (ClientResponse.Details.Error error : errors) {
+                            final Tuple failedTuple = metricPointsWithTuple.get(error.getDatapoint());
+                            if(failedTuple != null) {
+                                failedTuples.add(failedTuple);
+                            }
+                        }
+
+                        for (Tuple batchedTuple : batchHelper.getBatchTuples()) {
+                            if(failedTuples.contains(batchedTuple)) {
+                                collector.fail(batchedTuple);
+                            } else {
+                                collector.ack(batchedTuple);
+                            }
+                        }
+
+                    } else {
+                        throw new RuntimeException("Some of the metric points failed with details: " + errors);
+                    }
+                } else {
+                    LOG.debug("Acknowledging batched tuples");
+                    batchHelper.ack();
+                }
+                metricPointsWithTuple.clear();
+            }
+        } catch (Exception e) {
```

Comment:
should this catch be inside `batchHelper.shouldFlush()` since whole batch is failed.